### PR TITLE
Add health checks to pithos pod

### PR DESCRIPTION
**Pull request by UnderGreen**

### CODE REVIEW

File: cmd/pithosctl/healthz.go [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+func init() {
+	pithosctlCmd.AddCommand(healthzCmd)
+	healthzCmd.PersistentFlags().StringVar(&accessKey, "access-key", "", "S3 access key")
+	healthzCmd.PersistentFlags().StringVar(&secretAccessKey, "secret-access-key", "", "S3 secret key")
+	healthzCmd.PersistentFlags().StringVar(&endpoint, "endpoint", defaults.HealthzEndpoint, "S3 endpoint address")
+	healthzCmd.PersistentFlags().StringVar(&bucket, "bucket", defaults.HealthzBucket, "S3 bucket name")
+
+	if err := bindFlagEnv(healthzCmd.PersistentFlags()); err != nil {
+		log.WithError(err).Error("Failed to bind environment variables to flags.")
```
Comment (a-palchikov) [inline cmd/pithosctl/healthz.go:52]: Is this not supposed to be fatal? What happens when this fails - will the binary end up misconfigured?

File: cmd/pithosctl/healthz.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+	pithosHealthz, err = pithos.NewHealthz(*config, bucket)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if err := pithosHealthz.Prepare(); err != nil {
+		return trace.Wrap(err)
+	}
+
+	log.Info("Starting endpoint.")
```
Comment (a-palchikov) [inline cmd/pithosctl/healthz.go:73]:
```go
log.WithField("endpoint", endpoint).Info("Start server.")
```

File: cmd/pithosctl/main.go [hunk 3]
```diff
@@ -51,3 +59,19 @@ func init() {
...
+// bindFlagEnv binds environment variables to command flags
+func bindFlagEnv(flagSet *flag.FlagSet) error {
+	for env, flag := range envs {
+		cmdFlag := flagSet.Lookup(flag)
+		if cmdFlag != nil {
+			if value := os.Getenv(env); value != "" {
+				if err := cmdFlag.Value.Set(value); err != nil {
+					return trace.Wrap(err)
+				}
+
```
Comment (a-palchikov) [inline cmd/pithosctl/main.go:72]: nit: remove empty line?

File: internal/pithosctl/pkg/defaults/defaults.go [hunk 4]
```diff
@@ -25,4 +25,8 @@ const (
 	PithosNodeSelector = "pithos-role=node"
 	// Threshold is minimum bytes to consider load between nodes unevenly distributed
 	Threshold int64 = 1073741824 // 1GiB
+	// HealthzBucket is the name of the bucked used for health checks
```
Comment (a-palchikov) [inline internal/pithosctl/pkg/defaults/defaults.go:28]: bucke**t**
What kind of bucket is this?

File: internal/pithosctl/pkg/pithos/healthz.go [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+	insecure := false
+	client, err := minio.NewV2(config.Endpoint, config.AccessKey, config.SecretAccessKey, insecure)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return client, nil
+}
+
+// Prepare creates bucket and object if they are not exist
```
Comment (a-palchikov) [inline internal/pithosctl/pkg/pithos/healthz.go:96]: `if they do not exist`
Will it fail if the bucket/object exist?

File: internal/pithosctl/pkg/pithos/healthz.go [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+
+func (h *Healthz) bucketExists() (bool, error) {
+	found, err := h.Client.BucketExists(h.Bucket)
+	if err != nil {
+		return false, trace.Wrap(err)
+	}
+	return found, nil
+}
+
+func (h *Healthz) createBucket() error {
```
Comment (a-palchikov) [inline internal/pithosctl/pkg/pithos/healthz.go:117]: `createBucketIfNotExist`

File: resources/pithos.yaml [hunk 7]
```diff
@@ -153,6 +153,30 @@ spec:
...
+              secretKeyRef:
+                name: pithos-keys
+                key: master.secret
+        livenessProbe:
+          httpGet:
+            path: /healthz
+            port: 8080
+          initialDelaySeconds: 120
+          periodSeconds: 60
+          # a little bith higher than configured timeout in pithos.conf
```
Comment (a-palchikov) [inline resources/pithos.yaml:177]: bit~~h~~
`# slightly higher than the timeout configured in pithos.conf`
How do you know the timeout configured in pithos.conf is less than the hard-coded value in general?

**Comment by UnderGreen** [comment]
At first, I wanted server to serve both, but then I realized that server requires a lot more variables for `/status` check. The server part should not be used in pithos pod, because it does not make sense. About `:8080` you are right, I should change it to something different.

File: cmd/pithosctl/healthz.go [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+	"github.com/gravitational/pithos-app/internal/pithosctl/pkg/defaults"
+	"github.com/gravitational/pithos-app/internal/pithosctl/pkg/pithos"
+
+	"github.com/gravitational/trace"
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+)
+
+var (
+	healthzCmd = &cobra.Command{
```
[the remark this answers is not in the document]
Comment (UnderGreen) [reply cmd/pithosctl/healthz.go:31]: Maybe it is something in 1.13, which I am using on my machine.

File: cmd/pithosctl/healthz.go (hunk 1 above)
Comment (UnderGreen) [reply cmd/pithosctl/healthz.go:52]: Empty access/secret keys will fail later in CheckOrSetDefault. Empty bucket/endpoints will be set to defaults.

File: internal/pithosctl/pkg/pithos/healthz.go (hunk 5 above)
Comment (UnderGreen) [reply internal/pithosctl/pkg/pithos/healthz.go:96]: Did not check it in my testing. Will do it today.

Comment (UnderGreen) [reply internal/pithosctl/pkg/pithos/healthz.go:96]: In that case, PUT is used instead of POST, which makes Prepare funcion save to use on multiple pods.

**Pull request merged**